parse_xml.py

```diff
@@ -21,115 +22,207 @@ def changeExt(fname, ext):
+                    if line.strip().startswith("<?xml version"):
+                        continue
+                    elif line.strip().startswith("<!DOCTYPE"):
+                        continue
                     else:
-                       logging.info("-- Sending file: "+docid+" to Solr")
-                       response = sendToSolr('ptab', jsontext)
-                       r = response.json()
-                       status = r["responseHeader"]["status"]
-                       if status == 0:
-                           logfile.write(docid+"\n")
-                           logging.info("-- Solr update for file: "+docid+" complete")
-                       else:
-                           logging.info("-- Solr error for doc: "+docid+" error: "+', '.join("{!s}={!r}".format(k,v) for (k,v) in rdict.items()))
+                        filecontent.append(line)
+                else:
+                    filecontent.append("</main>")
+                outputFile(fn, filecontent)
+                logging.info("-- Combine file process complete")
+        else:
+            logging.info("-- File: "+fn+" already exists")
     except IOError as e:
-        logging.error("I/O error({0}): {1}".format(e.errno,e.strerror))
+        logging.error("-- I/O error({0}): {1}".format(e.errno,e.strerror))
+        raise
     except:
-        logging.error("Unexpected error:", sys.exc_info()[0])
+        logging.error("-- Unexpected error:", sys.exc_info()[0])
         raise
 
 #this function contains the code for parsing the xml file
 #and writing the results out to a json file
-def processXML(fname):
+def parseXML(fname):
     try:
-        with open(os.path.abspath(fname)) as fd:
-            fn = changeExt(fname,'json')
-            doc = xmltodict.parse(fd.read())
-            records = doc['main']['DATA_RECORD']
-            for x in records:
-                docid = x.get('DOCUMENT_IMAGE_ID',x.get('DOCUMENT_NM'))
-                txtfn = os.path.join(os.path.dirname(fn),'PDF_image',docid+'.txt')
-                if os.path.isfile(txtfn):
-                    x['LAST_MODIFIED_TS'] = formatDate(x, 'LAST_MODIFIED_TS')
-                    x['PATENT_ISSUE_DT'] = formatDate(x, 'PATENT_ISSUE_DT')
-                    x['DECISION_MAILED_DT'] = formatDate(x, 'DECISION_MAILED_DT')
-                    x['PRE_GRANT_PUBLICATION_DT'] = formatDate(x, 'PRE_GRANT_PUBLICATION_DT')
-                    x['APPLICANT_PUB_AUTHORIZATION_DT'] = formatDate(x, 'APPLICANT_PUB_AUTHORIZATION_DT')
-                    x['appid'] = x.pop('BD_PATENT_APPLICATION_NO')
-                    x['doc_date'] = x.pop('DOCUMENT_CREATE_DT')
-                    with open(txtfn) as dr:
-                        text = dr.read()
-                        x['textdata'] = text
+        fn = changeExt(fname,'json')
+        if not os.path.isfile(fn):
+            with open(fname) as fd:
+                logging.info("-- Beginning read of file at: "+str(datetime.now()))
```

Comment:
Prefer to just embed the timestamp in the logging format. Requires a little bit of setup. From the [Python Logging Cookbook][1]:

```
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
handler.setFormatter(formatter)
```

[1]: https://docs.python.org/3/howto/logging-cookbook.html